tensorflow/core/kernels/segment_reduction_ops_gpu.cu.cc

```diff
@@ -150,6 +154,422 @@ bool DisableSegmentReductionOpDeterminismExceptions() {
+                                Tinit>,
+      grid, block, shared_memory_bytes, d.stream(), nouter, ninner_vec,
+      nsegments, reduce_op, initial_value, empty_segment_value, is_mean,
+      is_sqrtn, input_vec, segment_offsets, indices, output_vec);
+}
+
+template <typename Tvec, typename Treducevec, typename Tindex,
+          typename Tsegmentids, typename Tinit>
+__global__ void SegmentReduceEpilogueKernel(
+    Tsegmentids nsegments, Tinit empty_segment_value, bool is_mean,
+    bool is_sqrtn,
+    const Treducevec* __restrict__ output_raw,   // [nsegments]
+    const Tindex* __restrict__ segment_offsets,  // [nsegments + 1]
+    Tvec* __restrict__ output) {                 // [nsegments]
+  GPU_1D_KERNEL_LOOP(seg, nsegments) {
+    Tindex segment_size = segment_offsets[seg + 1] - segment_offsets[seg];
+    Treducevec val = output_raw[seg];
+    if (segment_size == 0) {
+      // Empty segment.
+      val = Treducevec(empty_segment_value);
+    } else if (is_mean) {
+      val /= Treducevec(segment_size);
+    } else if (is_sqrtn) {
+      val /= Treducevec(
+          sqrt(typename RealTypeIfComplex<Tinit>::type(segment_size)));
+    }
+    // Cast from Treducevec to Tvec.
+    output[seg] = static_cast<Tvec>(val);
+  }
+}
+
+// Normalizes output_raw based on segment size and casts from Treducevec to
+// Tvec. If Tvec == Treducevec, this is safe to call with output_raw == output.
+template <typename Tvec, typename Treducevec, typename Tindex,
+          typename Tsegmentids, typename Tinit>
+Status LaunchSegmentReduceEpilogueKernel(
+    const GPUDevice& d, Tsegmentids nsegments, Tinit empty_segment_value,
+    bool is_mean, bool is_sqrtn,
+    const Treducevec* output_raw,   // [nsegments]
+    const Tindex* segment_offsets,  // [nsegments + 1]
+    Tvec* output) {                 // [nsegments]
+  GpuLaunchConfig config = GetGpuLaunchConfig(
+      nsegments, d,
+      &SegmentReduceEpilogueKernel<Tvec, Treducevec, Tindex, Tsegmentids,
+                                   Tinit>,
+      /*dynamic_shared_memory_size=*/0, /*block_size_limit=*/0);
+  return GpuLaunchKernel(
+      SegmentReduceEpilogueKernel<Tvec, Treducevec, Tindex, Tsegmentids, Tinit>,
+      config.block_count, config.thread_per_block, 0, d.stream(), nsegments,
+      empty_segment_value, is_mean, is_sqrtn, output_raw, segment_offsets,
+      output);
+}
+
+template <typename Tto>
+struct CastFunctor {
+  template <typename T>
+  __device__ Tto operator()(const T& val) const {
+    return static_cast<Tto>(val);
+  }
+};
+
+template <typename Treducevec, typename Tvec, typename Tindex,
+          typename Tsegmentids, typename ReduceOp, typename Tinit>
+Status SegmentReduceGPUImpl(
+    OpKernelContext* ctx, Tindex nouter, Tindex ninner_vec,
+    Tsegmentids nsegments, ReduceOp reduce_op, Tinit initial_value,
+    Tinit empty_segment_value, bool is_mean, bool is_sqrtn,
+    const Tvec* input_vec,           // [nouter or any, ninner_vec]
```

Comment:
What do you mean by "or any"?

Reply:
Added comment explaining. When `indices` is provided, the first dim of `input_vec` is indexed indirectly and has no specific size requirement (as long as it covers the values in `indices`).